tests/test_main.py

```diff
@@ -1755,27 +1751,25 @@ def test_final_field_decl_with_default_val(ann):
     class Model(BaseModel):
         a: ann = 10
 
-    Model.model_rebuild(ann=ann)
-
     assert 'a' in Model.__class_vars__
     assert 'a' not in Model.model_fields
 
 
-@pytest.mark.xfail(reason='waiting for https://github.com/pydantic/pydantic-core/pull/237')
 def test_final_field_reassignment():
     class Model(BaseModel):
+        model_config = ConfigDict(validate_assignment=True)
```

Comment:
This is a change from v1 — in v1:
* it would error for setting values to final fields even without having `validate_assignment=True` (or anything like it)
* it would raise a TypeError, not a ValidationError.

Do we need to change this? If so, I guess that would require a bit more work to support `Final` since it would presumably no longer be a pure pydantic_core thing. Is it possible to turn on `validate_assignment` on a per-field basis in pydantic core?

Reply:
For reference:

https://github.com/pydantic/pydantic/blob/11f7f9019016659e81c96481ffe5c42637489cf3/tests/test_main.py#L2283-L2293

Reply:
I think fine to keep the change.

Again, if they come with pitchforks, we can have special logic in core to raise a type error perhaps.

Reply:
I'm fine changing the type error, but I'm more concerned about needing to specify `validate_assignment=True` for it to be used. (Which might not be desirable.)